rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -141,26 +140,37 @@ class ReplayRSpace[F[_], C, P, E, A, R, K](store: IStore[F, C, P, A, K], branch:
               }
       } yield r
 
-//    @tailrec <- todo fix tailrec
     @SuppressWarnings(Array("org.wartremover.warts.Throw"))
     // TODO stop throwing exceptions
-    def getCommOrDataCandidates(comms: Seq[COMM]): F[Either[COMM, Seq[DataCandidate[C, R]]]] =
-      comms match {
-        case Nil =>
-          val msg = "List comms must not be empty"
-          logger.error(msg)
-          throw new IllegalArgumentException(msg)
-        case commRef :: Nil =>
-          runMatcher(commRef).map {
-            case Some(x) => Right(x)
-            case None    => Left(commRef)
-          }
-        case commRef :: rem =>
-          runMatcher(commRef).flatMap {
-            case Some(x) => Right(x).asInstanceOf[Either[COMM, Seq[DataCandidate[C, R]]]].pure[F]
-            case None    => getCommOrDataCandidates(rem)
-          }
+    def getCommOrDataCandidates(comms: Seq[COMM]): F[Either[COMM, Seq[DataCandidate[C, R]]]] = {
+      type COMMOrData  = Either[COMM, Seq[DataCandidate[C, R]]]
+      type Accumulator = (Seq[COMM], Option[COMMOrData])
+      implicit class RichAccumulator(acc: Accumulator) {
+        def comms: Seq[COMM]           = acc._1
+        def result: Option[COMMOrData] = acc._2
       }
+      object Accumulator {
+        def apply(comms: Seq[COMM], cp: Option[COMMOrData] = None): Accumulator = (comms, cp)
+      }
+      def go(acc: Accumulator) =
+        acc.comms match {
+          case Nil =>
+            val msg = "List comms must not be empty"
+            logger.error(msg)
+            throw new IllegalArgumentException(msg)
+          case commRef :: Nil =>
+            runMatcher(commRef).map {
+              case Some(x) => Accumulator(Nil, Some(x.asRight[COMM]))
+              case None    => Accumulator(Nil, Some(Left(commRef)))
+            }
+          case commRef :: rem =>
+            runMatcher(commRef).map {
+              case Some(x) => Accumulator(rem, Some(x.asRight[COMM]))
+              case None    => Accumulator(rem, None)
+            }
+        }
+      concurrent.iterateWhileM(Accumulator(comms))(go)(_.result.isEmpty).map(_.result.get)
```

Comment:
this is a nice improvement, but wouldn't it be clearer to use

```scala
def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B]
```

directly?

I think
```scala
type Accumulator = (Seq[COMM], Option[COMMOrData])

//because of
// (A, Option[B]) === Either[A, (A, B)]

//is isomorphic to 
type Accumulator = Either[Seq[COMM], (Seq[COMM], COMMOrData)]  
```

which fits nicely the tailrecM signature.

It also seems to me that we only need the `Seq[COMM]` during iteration, so we could
tailrecM with `f: Seq[COMM] => Either[Seq[COMM], COMMOrData]` and get the COMMOrData that we need on the output without any `.get`-s or `_._2`-s

Reply:
and we'd get rid of `Option.get` (yay!)

Reply:
Aaand we'd not need the Accumulator type anymore

Reply:
Unless I'm wrong ofc